src/merkle/new_smt/mod.rs

```diff
@@ -0,0 +1,272 @@
+use core::cmp::Ordering;
+
+use winter_math::StarkField;
+
+use crate::hash::rpo::Rpo256;
+use crate::utils::{collections::Vec, vec};
+use crate::{Felt, EMPTY_WORD};
+
+use super::{
+    BTreeMap, BTreeSet, EmptySubtreeRoots, InnerNode, LeafIndex, MerkleError, NodeIndex, RpoDigest,
+    SparseMerkleTree, Word,
+};
+
+#[cfg(test)]
+mod tests;
+
+// CONSTANTS
+// ================================================================================================
+
+pub const NEW_SMT_DEPTH: u8 = 64;
+
+// SMT
+// ================================================================================================
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct NewSmt {
+    root: RpoDigest,
+    leaves: BTreeMap<u64, NewSmtLeaf>,
+    inner_nodes: BTreeMap<NodeIndex, InnerNode>,
+}
+
+impl NewSmt {
+    /// Returns a new [NewSmt].
+    ///
+    /// All leaves in the returned tree are set to [ZERO; 4].
+    pub fn new() -> Self {
+        let root = *EmptySubtreeRoots::entry(NEW_SMT_DEPTH, 0);
+
+        Self {
+            root,
+            leaves: BTreeMap::new(),
+            inner_nodes: BTreeMap::new(),
+        }
+    }
+
+    /// Returns a new [SimpleSmt] instantiated with leaves set as specified by the provided entries.
+    ///
+    /// All leaves omitted from the entries list are set to [ZERO; 4].
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The number of entries exceeds 2^63 entries.
+    /// - The provided entries contain multiple values for the same key.
+    pub fn with_entries(
+        entries: impl IntoIterator<Item = (NewSmtKey, Word)>,
+    ) -> Result<Self, MerkleError> {
+        // create an empty tree
+        let mut tree = Self::new();
+
+        // compute the max number of entries. We use an upper bound of depth 63 because we consider
+        // passing in a vector of size 2^64 infeasible.
+        let max_num_entries = 2_usize.pow(tree.depth().min(63).into());
+
+        // This being a sparse data structure, the EMPTY_WORD is not assigned to the `BTreeMap`, so
+        // entries with the empty value need additional tracking.
+        let mut key_set_to_zero = BTreeSet::new();
+
+        for (idx, (key, value)) in entries.into_iter().enumerate() {
+            if idx >= max_num_entries {
+                return Err(MerkleError::InvalidNumEntries(max_num_entries));
+            }
+
+            let old_value = tree.update_leaf(key, value);
+
+            if old_value != EMPTY_WORD || key_set_to_zero.contains(&key) {
+                return Err(MerkleError::DuplicateValuesForIndex(
+                    LeafIndex::<NEW_SMT_DEPTH>::from(key).value(),
+                ));
```

Comment:
This is not great; I should add a new variant `MerkleError::DuplicateValuesForKey(NewSmtKey)`, which bloats `MerkleError` even more, highlighting the need for a `MerkleError` refactor.

I don't think we should have a one-size-fits-all `MerkleError`. Instead, every data structure should use return its own `Error`. Even, every method should return an `Error` type that contains only variants for errors that can actually occur in it. These method-level errors can be aggregated into a larger `TreeTypeError` if needed. This is made clean by using `thiserror` - specifically the `#[from]` attribute.

This PR is getting big though, and we should do this in another PR (and agree on the right strategy in an issue).

Reply:
Yep - let's create an issue for this.

Reply:
#247